mmdet/core/bbox/assigners/hungarian_assigner.py

```diff
@@ -0,0 +1,156 @@
+import torch
+from scipy.optimize import linear_sum_assignment
+
+from ..builder import BBOX_ASSIGNERS
+from ..iou_calculators import build_iou_calculator
+from ..transforms import bbox_cxcywh_to_xyxy, bbox_xyxy_to_cxcywh
+from .assign_result import AssignResult
+from .base_assigner import BaseAssigner
+
+
+@BBOX_ASSIGNERS.register_module()
+class HungarianAssigner(BaseAssigner):
+    """Computes one-to-one matching between predictions and ground truth.
+
+    This class computes an assignment between the targets and the predictions
+    based on the costs. The costs are weighted sum of three components:
+    classfication cost, regression L1 cost and regression giou cost. The
+    targets don't include the no_object, so generally there are more
+    predictions than targets. After the one-to-one matching, the un-matched
+    are treated as backgrounds. Thus each query prediction will be assigned
+    with `0` or a positive integer indicating the ground truth index:
+
+    - 0: negative sample, no assigned gt
+    - positive integer: positive sample, index (1-based) of assigned gt
+
+    Args:
+        cls_weight (int | float, optional): The scale factor for classification
+            cost. Default 1.0.
+        bbox_weight (int | float, optional): The scale factor for regression
+            L1 cost. Default 1.0.
+        iou_weight (int | float, optional): The scale factor for regression
+            iou cost. Default 1.0.
+        mode (str, optional): "iou" (intersection over union), "iof"
+            (intersection over foreground), or "giou" (generalized intersection
+            over union). Default "giou".
+    """
+
+    def __init__(self,
+                 cls_weight=1.,
+                 bbox_weight=1.,
+                 iou_weight=1.,
+                 iou_calculator=dict(type='BboxOverlaps2D'),
+                 mode='giou'):
+        assert mode in ['iou', 'iof', 'giou'], f'Unsupported mode {mode}.'
```

Comment:
1. We should not do the check here. Move the assertion to iou_calculator.
2. make `mode='giou'` an argument of iou_calculator, it should not be the argument of the assigner.